src/app/rmmapi/rbwebmail.ts

```diff
@@ -222,14 +222,62 @@ export class RunboxWebmailAPI {
                 messageContentsObservable.next(r);
                 messageContentsObservable.complete();
             }, err => {
-                delete this.messageContentsCache[messageId];
                 messageContentsObservable.error(err);
+                delete this.messageContentsCache[messageId];
             });
 
             return messageContentsObservable;
         }
     }
 
+    public downloadMessages(messageIds: number[]): Promise<MessageContents[]> {
+        const missingMessages = [];
+        for (const msgid of messageIds) {
+            if (!this.messageContentsCache[msgid]) {
+                this.messageContentsCache[msgid] = new AsyncSubject<MessageContents>();
+                missingMessages.push(msgid);
+            }
+        }
+
+        const messagePromises = messageIds.map(id => this.messageContentsCache[id].toPromise());
+
+        if (missingMessages.length > 0) {
+            this.http.get(`/rest/v1/email/download/${missingMessages.join(',')}`).pipe(
+                catchError((err: HttpErrorResponse) => throwError(new Error(err.message))),
+                concatMap((res: any) => {
+                    if (res.status === 'success') {
+                        return of(res.result);
+                    } else {
+                        return throwError(new Error(res.errors[0]));
+                    }
+                }),
+            ).subscribe(
+                (result: any) => {
+                    for (const resultKey in result) {
+                        const msgid = parseInt(resultKey);
+                        const contents = result[msgid]?.json;
+                        if (contents) {
+                            this.messageContentsCache[msgid].next(contents);
+                            this.messageContentsCache[msgid].complete();
+                        } else {
+                            this.messageContentsCache[msgid].error(result[msgid]?.error);
+                            delete this.messageContentsCache[msgid];
+                        }
+                    }
+                },
+                (err: Error) => {
+                    for (const msgid of missingMessages) {
+                        this.messageContentsCache[msgid].error(err.toString());
+                        delete this.messageContentsCache[msgid];
+                    }
+                }
+            );
+        }
+
+        //return Promise.allSettled(messagePromises);
```

Comment:
Not sure yet if we'd actually prefer `allSettled()` here or not, but it doesn't compile under our current setup so far anyway ;) If we go for it, we might need a polyfill to make it work on ancient browsers too.